baseline-error-prone/src/main/java/com/palantir/baseline/errorprone/DuplicateArgumentTypes.java

```diff
@@ -0,0 +1,115 @@
+/*
+ * (c) Copyright 2020 Palantir Technologies Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.palantir.baseline.errorprone;
+
+import com.google.auto.service.AutoService;
+import com.google.common.collect.Streams;
+import com.google.errorprone.BugPattern;
+import com.google.errorprone.VisitorState;
+import com.google.errorprone.bugpatterns.BugChecker;
+import com.google.errorprone.matchers.Description;
+import com.google.errorprone.matchers.Matchers;
+import com.google.errorprone.suppliers.Suppliers;
+import com.google.errorprone.util.ASTHelpers;
+import com.sun.source.tree.MethodTree;
+import com.sun.source.tree.Tree;
+import com.sun.tools.javac.code.Type;
+import java.util.concurrent.atomic.AtomicBoolean;
+
+/**
+ * Development Practices: Writing good unit tests.
+ */
+@AutoService(BugChecker.class)
+@BugPattern(
+        name = "DuplicateArgumentTypes",
+        link = "https://github.com/palantir/gradle-baseline#baseline-error-prone-checks",
+        linkType = BugPattern.LinkType.CUSTOM,
+        severity = BugPattern.SeverityLevel.SUGGESTION,
+        summary = "Duplicate argument types")
+public final class DuplicateArgumentTypes extends BugChecker implements BugChecker.MethodTreeMatcher {
+
+    @Override
+    public Description matchMethod(MethodTree tree, VisitorState state) {
+        AtomicBoolean badMethod = new AtomicBoolean(false);
+
+        tree.getParameters().forEach(param1 -> {
+            Tree outerTree = param1.getType();
+            tree.getParameters().forEach(param2 -> {
+                if (!param1.equals(param2)) {
+                    Tree innerTree = param2.getType();
+                    badMethod.set(badMethod.get() || isSubTypeOf(outerTree, innerTree, state));
+                }
+            });
+        });
+
+        if (badMethod.get()) {
+            return buildDescription(tree)
+                    .setMessage("Some argument types are equal or are subtypes of each other. Consider using a "
+                            + "builder instead.")
+                    .build();
+        }
+
+        return Description.NO_MATCH;
+    }
+
+    private boolean isSubTypeOf(Tree tree1, Tree tree2, VisitorState state) {
```

Comment:
this method itself would get flagged by your check :)

Reply:
I know - I should use a builder, but the way this code is written, it doesn't matter about the ordering. Still, I should practice what I preach!

Reply:
I was kinda thinking that we could initially have this for publicly visible methods.